### DISCUSSION TITLE
Track metric events before host is started (.NET / ASP.NET Core)

### DISCUSSION THREAD

**Discussion by Skyppid** [getsentry/sentry-dotnet#3409] [Q&A]
Well we currently have a service framework implemented which allows us to run maintenance procedures before the service goes live (hosting is started with `app.Run(Async)` and accepts requests.

We want to trace these procedures as they can be quite long-running to measure how much time they usually take and if timing gets worse over time (likely due to more data in the system) or better (due to performance improvements).

Problem: Sentry seems to start with the `app.Run` call when the Host actually starts and seemingly all `SentrySdk.Metrics.*` calls are lost that happen before that.

Is there either a way to start Sentry earlier or is there some kind of caching feature that can track these events and send them once Sentry is up and running?

It would be a bummer if that wasn't possible. Moving the maintenance routines after hosting started would be bad as they are required before any requests are accepted.

Thanks for the help :)

**Comment by jamescrosswell** [comment]
If it's just timings you're after, you could easily cache these and do something like:

```csharp
MyTiming timing = new MyTiming("some-task");
// Do some maintenance stuff
timing.Finish();

app.Run();
...
SentrySdk.Metrics.Timing(
    timing.Key,
    timing.Stopwatch.Elapsed.TotalMilliseconds,
    MeasurementUnit.Duration.Millisecond,
    tags: timing.Tags,
    timestamp: timing.StartTime
    );

public class MyTiming
{
    public string Key { get; }
    public readonly DateTime StartTime;
    public readonly Stopwatch Stopwatch = new();
    public IDictionary<string, string>? Tags;

    public MyTiming(string key, params KeyValuePair<string, string>[] tags)
    {
        Key = key;
        Tags = tags.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
        StartTime = DateTime.UtcNow;
        Stopwatch.Start();
    }
    
    public void Finish()
    {
        Stopwatch.Stop();
    }
}
```

By passing in a `timestamp` argument to the `Metrics.Timing` method, you're essentially backdating the timing.

**Reply by Skyppid** [reply]
Ah thanks for the tip. But after hours they actually showed up in Sentry, so I guess Sentry already does cache them internally and send them out as soon as possible. Wasn't really clear from the logs though, but since the data is there I guess it's not a problem after all. Might rather be related to Metrics being a beta feature I guess.

Still a good tip to consider :)

**Reply by bitsandfoxes** [reply]
> But after hours they actually showed up in Sentry

This might be an issue how they are getting displayed in the dashboard. Have you tried reducing the timeframe? I noticed that when I "zoom" down to somewhere in the last hour / last 5 minutes that they show way earlier.

**Reply by Skyppid** [reply]
@bitsandfoxes nope, but thanks for the tip, I'll try that next time :)
